src/goto-programs/goto_program.cpp

```diff
@@ -532,10 +533,36 @@ std::string as_string(
 /// order they appear in the goto program.
 void goto_programt::compute_loop_numbers()
 {
-  unsigned nr=0;
+  // This deals with nested loop number assignment by treating the queue
```

Comment:
This needs some testing regarding what happens when abnormally shaped GOTO programs are used -- e.g. entry in the middle of a loop:
```
// Enter the loop:
goto entry;

head:
  ...
entry:
  ...
goto head;
```

Or a loop with multiple backedges (more than one `goto head;` statement), or a loop with multiple edges:

```
if(x)
  goto entry1;
else
  goto entry2;

entry1:
  if(iterate())
    goto entry2;

entry2:
  if(iterate())
    goto entry1;
```

We might be able to get by with a hackish loop characterisation or might need the real natural-loops pass, but either way we should have a go at breaking it and check something halfway sensible happens.